plutus-chain-index/src/Plutus/ChainIndex/Lib.hs

```diff
@@ -170,26 +171,16 @@ getTipSlot config = do
     }
   pure slotNo
 
-showProgress :: IORef Integer -> Slot -> Slot -> IO ()
-showProgress lastProgressRef (Slot tipSlot) (Slot blockSlot) = do
-    lastProgress <- readIORef lastProgressRef
-    let pct = (100 * blockSlot) `div` tipSlot
-    if pct > lastProgress then do
-        putStrLn $ "Syncing (" ++ show pct ++ "%)"
-        writeIORef lastProgressRef pct
-    else pure ()
-
--- | Changes the given @ChainSyncHandler@ to print out the synchronisation progress percentages.
-showingProgress :: ChainSyncHandler -> IO ChainSyncHandler
-showingProgress handler = do
-    -- The primary purpose of this query is to get the first response of the node for potential errors before opening the DB and starting the chain index.
-    -- See #69.
-    lastProgressRef <- newIORef 0
+-- | Write 'ChainSyncEvent's from the 'ChainSyncHandler' to the 'TChan'.
+writeChainSyncEventToChan
+    :: TChan ChainSyncEvent
+    -> ChainSyncHandler
+    -> IO ChainSyncHandler
+writeChainSyncEventToChan tchan handler = do
     pure $ \case
-        (RollForward block@(CI.Block (CI.Tip blockSlot _ _) _) tip@(CI.Tip tipSlot _ _)) -> do
-            showProgress lastProgressRef tipSlot blockSlot
-            handler $ RollForward block tip
-        evt -> handler evt
+        evt -> do
+            atomically $ writeTChan tchan evt
```

Comment:
These events can be pretty big I think? Dumping everything in a channel for 30s doesn't sound like a good idea. I guess you'd want a function like

```haskell
writeToChan
     :: (ChainSyncEvent -> a)
     -> TChan a
     -> ChainSyncHandler
     -> IO ChainSyncHandler
```

so you can only keep in memory what you need.

Reply:
Right! :)